youtube_dl/extractor/iqm2.py

```diff
@@ -0,0 +1,109 @@
+# JWPlayer JS loader" (line 2372 as of 6 Oct 2016).
+
+# It appears that the metadata associated with the video (like its
+# title) does not appear anywhere in the 2 HTML pages that get
+# downloaded through this extractor. So it would need to download
+# additional HTTP resources in order to get "real" metadata.
+
+# This also appears to be the only example to date of an extractor
+# that calls-out to the generic extractor, so it may be
+# useful as an example. Or perhaps it means that there's a better way
+# to do this and it should be rewritten differently, esp. to not
+# leverage the generic? (xxx)
+
+# Contributed by John Hawkinson <jhawk@mit.edu>, 6 Oct 2016.
+
+# Potential test URLs:
+# http://cambridgema.iqm2.com/Citizens/SplitView.aspx?Mode=Video&MeetingID=1679
+# https://CambridgeMA.IQM2.com/Citizens/VideoMain.aspx?MeetingID=1679
+
+class IQM2IE(InfoExtractor):
+
+    # xxx is really right that InfoExtractor.suitable() calls re.compile()
+    # on _VALID_URL in a case-sensitive fashion? It's obviously reasonable
+    # for the path portion of a URL to be case-sensitive, but the hostname
+    # ought not to be. And it seems like strict adherence might mess up a
+    # bunch of extractors in funny-cased URLs? Redefine suitable() to search
+    # case-insensitively. Note this also changes the re.match() call at the
+    # start of _real_extract()
+    #
+    # In this case, we commonly see both iqm2.com and IQM2.com.
+    
+    @classmethod
+    def suitable(cls, url):
+        """Receives a URL and returns True if suitable for this IE."""
+
+        # This does not use has/getattr intentionally - we want to know whether
+        # we have cached the regexp for *this* class, whereas getattr would also
+        # match the superclass
+        if '_VALID_URL_RE' not in cls.__dict__:
+            cls._VALID_URL_RE = re.compile(cls._VALID_URL, flags=re.IGNORECASE)
+        return cls._VALID_URL_RE.match(url) is not None
+
+    _VALID_URL = r'https?://(?:\w+\.)?iqm2\.com/Citizens/\w+.aspx\?.*MeetingID=(?P<id>[0-9]+)'
+    _TEST = {
+        'url': 'http://cambridgema.iqm2.com/Citizens/SplitView.aspx?Mode=Video&MeetingID=1679#',
+        'md5': '478ea30eee1966f7be0d8dd623122148',
+        'info_dict': {
+            'id': '1563_720',
+            'ext': 'mp4',
+            'title': 'Cambridge, MA (2)',
+            'uploader': 'cambridgema.iqm2.com',
+        }
+    }
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url, flags=re.IGNORECASE)
+        video_id = mobj.group('id')
+
+        webpage = self._download_webpage(url, video_id)
+
+        # print "Original URL is", url
+
+        # We want to extract an inner URL like this:
+        # <div id="VideoPanel" class="LeftTopContent">
+        #   <div id="VideoPanelInner" onselectstart="javascript:return false;"
+        #     style="overflow:hidden"
+        #     src="/Citizens/VideoScreen.aspx?MediaID=1563&Frame=SplitView"
+        #     quality="hd">
```

Comment:
I think we care whether the maintainers of the code can read the regexp. What is the cost to having such a comment, it would seem to only benefit?

I think your argument cuts the other way as well. If the HTML has in fact changed, then someone looking to update the extractor may have a really tough time trying to figure out what the regexp was supposed to do, because they inspect the HTML and find it doesn't seem to be close to relevant. Having the example of the regexp makes it that much easier to see what kind of change was there, and to make the code more understandabe.

I'm certainly not going to die on this one, but I strongly believe more comments make things clearer, and that regexps are so bad for intelligibility that it's worth a little time to improve clarity.

Douglas Crockford has written a bit about this problem; in his book _Javascript: The Good Parts_, in Chapter 7, he wrote:

> Regular expressions tend to be extremely terse, even cryptic. They are easy to use in their simplest form, but they can quickly become bewildering. JavaScript's regular expressions are difficult to read in part because they do not allow comments or whitespace. All of the parts of a regular expression are pushed tightly together, making them almost indecipherable. This is a particular concern when they are used in security applications for scanning and validation. If you cannot read and understand a regular expression, how can you have confidence that it will work correctly for all inputs? Yet, despite their obvious drawbacks, regular expressions are widely used.

Now, of course, python's regexps are a bit better since they support (?x) and it looks like about 1/6 of the extractors do so. So at least there's the oportunity to write better regexps. But my takeaway is that one good way to mitigate the crypticness and errorprone nature of regexps is to try to use them as simply and clearly as possible, and that means showing a reader what they mean.

Because again, what's the cost? 

Obviously nobody's going to get confused by this particular regexp, which is about as boring as they come in youtube-dl _VALID_URL contexts. But I'm writing for the general principle...